tools/clang/lib/CodeGen/CGHLSLMS.cpp

```diff
@@ -1921,6 +1940,52 @@ void CGMSHLSLRuntime::AddHLSLFunctionInfo(Function *F, const FunctionDecl *FD) {
   }
 }
 
+void CGMSHLSLRuntime::RemapObsoleteSemantic(SourceLocation SLoc, DxilParameterAnnotation &paramInfo, bool isPatchConstantFunction) {
```

Comment:
Why is it necessary to remap in both CodeGen and HLSignatureLower?  I thought HLSignatureLower should catch all cases, while CodeGen could only catch top-level parameter semantics.  If the information is used before HLSignatureLower, then only looking at top-level parameter semantics in CodeGen only works in some cases and not others.  Also, if both places are necessary, it would seem better to share code if possible rather than duplicate it.

Reply:
The semantics that are specified directly on the entry function return value or its parameters are validated during CodeGen. However, if the entry function uses a struct with annotated members, these are not validated until later in HLSignatureLower. I did not put it in a shared code because of slightly different inputs and error handling but I will try again to see if I can make it work.